### PR TITLE
feat: add load testing harness, coder loadtest command

### PR DESCRIPTION
**Pull request by deansheather**
- Adds `httpapi.Duration` which wraps `time.Duration` with sensible JSON marshaling and unmarshaling
- Adds `loadtest/harness` package with tests which is the overall runner of a collection of tests
    - Concurrency can be controlled by using a different execution strategy
- Adds `loadtest/placebo` which has an implementation of `harness.Runnable` that sleeps for a given duration with jitter and failure chance
- Adds `loadtest/workspacebuild` which has an implementation of `harness.Runnable` that builds a workspace, waits for it to become available
    - Also implements `harness.Cleanable` to delete the workspace on test completion
- Adds hidden `coder loadtest` subcommand which is only available to the site owner
    - Configured by a JSON file (accepts STDIN as well)
    - Runs any variety of tests with any given concurrency settings
    - Prints error message and logs for all failed test runs
    - Cleans up afterwards

## TODO (in future PRs):
- Add a loadtest for workspace connections via STUN or DERP
- Better output formatting, including HTML and JSON output
- Live progress/failure output instead of only outputting information at the end
- Github actions workflows for periodic load testing in different configurations
- Documentation, json schema and unhidden loadtest command

<details>
  <summary><h2>Example</h2></summary>

```json
{
  "strategy": {
    "type": "concurrent"
  },
  "tests": [
    {
      "type": "placebo",
      "count": 500,
      "placebo": {
        "sleep": "1s",
        "failure_chance": 0.005
      }
    }
  ],
  "timeout": "1m"
}

```

```console
$ coder loadtest --config ./loadtest.config.json
Running load test...

== FAIL: placebo-0/322

        Error: test failed due to configured failure chance

        Log:
                sleeping for 1s
                failure chance is 0.005000
                rolling the dice of fate...
                rolled: 0.001123
                :(


Test results:
        Pass:  499
        Fail:  1
        Total: 500

        Total duration: 5.003371453s
        Avg. duration:  1.000593179s

Cleaning up...
```
</detail>

### CODE REVIEW

File: cli/loadtest.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+			th := harness.NewTestHarness(strategy)
+
+			for i, t := range config.Tests {
+				name := fmt.Sprintf("%s-%d", t.Type, i)
+
+				for i := 0; i < t.Count; i++ {
+					id := strconv.Itoa(i)
+					runner, err := t.NewRunner(client)
+					if err != nil {
+						return xerrors.Errorf("create %q runner for %s/%s: %w", t.Type, name, id, err)
```
Comment (coadler) [inline cli/loadtest.go:107]: The name string already has the id in it right? Seems redundant to include it again

File: cli/root.go [hunk 2]
```diff
@@ -70,33 +70,35 @@ func init() {
 }
 
 func Core() []*cobra.Command {
+	// Please re-sort this list alphabetically if you change it!
```
Comment (coadler) [inline cli/root.go:73]: true

File: coderd/httpapi/json.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import (
+	"encoding/json"
+	"time"
+
+	"golang.org/x/xerrors"
+)
+
+// Duration wraps time.Duration and provides better JSON marshaling and
+// unmarshaling.
+type Duration time.Duration
```
Comment (coadler) [inline coderd/httpapi/json.go:12]: Would be helpful to include what specifically is better about it.

**Review by coadler — APPROVED** [review]

File: cli/loadtest.go (hunk 1 above)
Comment (deansheather) [reply cli/loadtest.go:107]: The name is the name of the test e.g. `workspacebuild-0` and the id is the iteration of the test e.g. 0, 1, 2, 3, etc.

I reused the `i` variable in the two loops which makes this hard to read, so I've changed the inner loop to `j` to make it easier to read

Comment (coadler) [reply cli/loadtest.go:103]: ah

**Pull request merged**